File: android/app/src/main/kotlin/com/musicmuni/vozos/demo/sonix/simplified/MetronomeSectionSimplified.kt
package com.musicmuni.vozos.demo.sonix.simplified

import android.content.Context
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.musicmuni.vozos.sonix.SonixMetronome
import com.musicmuni.vozos.sonix.sample.components.OptionChip
import com.musicmuni.vozos.demo.util.copyAssetToFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Simplified Metronome Section using SonixMetronome.create().
 *
 * Demonstrates the zero-config factory pattern for metronome playback.
 * Compare this to MetronomeSection.kt which uses SonixMetronome.Builder().
 */
@Composable
fun MetronomeSectionSimplified(context: Context) {
    val scope = rememberCoroutineScope()

    var metronome by remember { mutableStateOf<SonixMetronome?>(null) }
    var bpm by remember { mutableFloatStateOf(120f) }
    var beatsPerCycle by remember { mutableIntStateOf(4) }
    var volume by remember { mutableFloatStateOf(0.8f) }
    var status by remember { mutableStateOf("Initializing...") }

    // Paths for reinitialization
    var samaSamplePath by remember { mutableStateOf<String?>(null) }
    var beatSamplePath by remember { mutableStateOf<String?>(null) }

    // UI state from SonixMetronome StateFlows
    val currentBeat by metronome?.currentBeat?.collectAsState() ?: remember { mutableStateOf(0) }
    val isRunning by metronome?.isPlaying?.collectAsState() ?: remember { mutableStateOf(false) }
    val isInitialized by metronome?.isInitialized?.collectAsState() ?: remember { mutableStateOf(false) }

    // Update status when initialized
    LaunchedEffect(isInitialized) {
        if (isInitialized) {
            status = "Ready"
        }
    }

    // Initialize metronome
    LaunchedEffect(Unit) {
        val (samaPath, beatPath) = withContext(Dispatchers.IO) {
            val sama = copyAssetToFile(context, "sama_click.wav")
            val beat = copyAssetToFile(context, "beat_click.wav")
            Pair(sama.absolutePath, beat.absolutePath)
        }
        samaSamplePath = samaPath
        beatSamplePath = beatPath

        // CREATE METRONOME - ZERO-CONFIG FACTORY!
        metronome = SonixMetronome.create(
            samaSamplePath = samaPath,
            beatSamplePath = beatPath,
            bpm = bpm,
            beatsPerCycle = beatsPerCycle
        )
        // Note: status will update to "Ready" via the isInitialized LaunchedEffect
    }

    // Cleanup
    DisposableEffect(Unit) {
        onDispose { metronome?.release() }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text("Metronome (Simplified API)", style = MaterialTheme.typography.titleMedium)
        Text("Status: $status", style = MaterialTheme.typography.bodySmall)

        // BPM slider
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("BPM:", modifier = Modifier.width(40.dp))
            Slider(
                value = (bpm - 60f) / 140f,
                onValueChange = { bpm = 60f + (it * 140f); metronome?.setBpm(bpm) },
                modifier = Modifier.weight(1f)
            )
            Text("${bpm.toInt()}", modifier = Modifier.width(40.dp))
        }

        // Beats per cycle
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("Beats:", modifier = Modifier.width(50.dp))
            listOf(3, 4, 6, 8).forEach { count ->
                OptionChip(
                    selected = beatsPerCycle == count,
                    onClick = {
                        if (!isRunning && beatsPerCycle != count) {
                            beatsPerCycle = count
                            // Reinitialize with new beat count
                            val samaPath = samaSamplePath
                            val beatPath = beatSamplePath
                            if (samaPath != null && beatPath != null) {
                                scope.launch {
                                    metronome?.release()
                                    metronome = SonixMetronome.create(
                                        samaSamplePath = samaPath,
                                        beatSamplePath = beatPath,
                                        bpm = bpm,
                                        beatsPerCycle = count
                                    )
                                    status = "Beats: $count"
                                }
                            }
                        }
                    },
                    label = count.toString(),
                    enabled = !isRunning
                )
            }
        }

        // Volume
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("Vol:", modifier = Modifier.width(40.dp))
            Slider(
                value = volume,
                onValueChange = { volume = it; metronome?.let { m -> m.volume = it } },
                modifier = Modifier.weight(1f)
            )
            Text("${(volume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Beat indicators
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text("Beat:", modifier = Modifier.width(50.dp))
            (0 until beatsPerCycle).forEach { beat ->
                val isCurrentBeat = beat == currentBeat && isRunning
                val isSama = beat == 0
                Box(
                    modifier = Modifier
                        .size(24.dp)
                        .clip(CircleShape)
                        .background(
                            when {
                                isCurrentBeat && isSama -> Color.Red
                                isCurrentBeat -> Color.Green
                                isSama -> Color.Red.copy(alpha = 0.3f)
                                else -> Color.Gray.copy(alpha = 0.3f)
                            }
                        )
                )
            }
        }

        // Control buttons
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(
                onClick = { metronome?.start(); status = "Running" },
                enabled = isInitialized && !isRunning
            ) { Text("Start") }
            Button(
                onClick = { metronome?.stop(); status = "Stopped" },
                enabled = isRunning
            ) { Text("Stop") }
        }
    }
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Sonix/MetronomeSectionSimplified.swift
import SwiftUI
import VoxaTrace

/// Simplified Metronome Section using SonixMetronome.create().
///
/// Demonstrates the zero-config factory pattern for metronome playback.
/// Uses type-safe observers (no force casts like `as! Bool`).
/// Compare this to MetronomeSection.swift which uses SonixMetronome.Builder().
struct MetronomeSectionSimplified: View {
    @State private var metronome: SonixMetronome?
    @State private var isRunning = false
    @State private var bpm: Float = 120
    @State private var beatsPerCycle: Int32 = 4
    @State private var currentBeat: Int32 = 0
    @State private var volume: Float = 0.8
    @State private var status = "Initializing..."
    @State private var isInitialized = false

    // Task handles for cancellation
    @State private var beatTask: Task<Void, Never>?
    @State private var playingTask: Task<Void, Never>?
    @State private var initializedTask: Task<Void, Never>?

    // Store paths for reinitialization
    @State private var samaSamplePath: String?
    @State private var beatSamplePath: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("Metronome (Simplified API)")
                    .font(.headline)
                Spacer()
                if !isInitialized {
                    ProgressView()
                        .scaleEffect(0.8)
                }
            }

            Text("Status: \(status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // BPM slider
            HStack {
                Text("BPM:")
                    .frame(width: 40, alignment: .leading)
                Slider(value: Binding(
                    get: { Double(bpm - 60) / 140 },
                    set: { newValue in
                        bpm = Float(60 + newValue * 140)
                        metronome?.setBpm(bpm: bpm)
                    }
                ))
                Text("\(Int(bpm))")
                    .frame(width: 40)
            }

            // Beats per cycle
            HStack {
                Text("Beats:")
                    .frame(width: 50, alignment: .leading)
                beatsButton(count: 3)
                beatsButton(count: 4)
                beatsButton(count: 6)
                beatsButton(count: 8)
            }

            // Volume
            HStack {
                Text("Vol:")
                    .frame(width: 40, alignment: .leading)
                Slider(value: Binding(
                    get: { Double(volume) },
                    set: { newValue in
                        volume = Float(newValue)
                        metronome?.volume = volume
                    }
                ))
                Text("\(Int(volume * 100))%")
                    .frame(width: 40)
            }

            // Beat indicators
            HStack {
                Text("Beat:")
                    .frame(width: 50, alignment: .leading)
                ForEach(0..<Int(beatsPerCycle), id: \.self) { beat in
                    let isCurrentBeat = beat == Int(currentBeat) && isRunning
                    let isSama = beat == 0
                    Circle()
                        .fill(beatColor(isCurrentBeat: isCurrentBeat, isSama: isSama))
                        .frame(width: 24, height: 24)
                }
            }

            // Control buttons
            HStack(spacing: 12) {
                Button("Start") {
                    metronome?.start()
                    status = "Running"
                }
                .disabled(!isInitialized || isRunning)
                .buttonStyle(.borderedProminent)

                Button("Stop") {
                    metronome?.stop()
                    status = "Stopped"
                    currentBeat = 0
                }
                .disabled(!isRunning)
                .buttonStyle(.bordered)
            }
        }
        .task {
            await initializeMetronome()
        }
        .onDisappear {
            // Cancel all observer tasks
            beatTask?.cancel()
            playingTask?.cancel()
            initializedTask?.cancel()
            metronome?.release()
        }
    }

    private func initializeMetronome() async {
        guard let samaPath = copyAssetToFile(name: "sama_click", ext: "wav"),
              let beatPath = copyAssetToFile(name: "beat_click", ext: "wav") else {
            await MainActor.run {
                status = "Click samples not found"
            }
            return
        }

        // Store paths for reinitialization
        samaSamplePath = samaPath
        beatSamplePath = beatPath

        // CREATE METRONOME - ZERO-CONFIG FACTORY!
        let newMetronome = SonixMetronome.create(
            samaSamplePath: samaPath,
            beatSamplePath: beatPath,
            bpm: bpm,
            beatsPerCycle: beatsPerCycle
        )

        await MainActor.run {
            metronome = newMetronome
            status = "Initializing audio..."

            // Type-safe observers - no force casts needed!
            initializedTask = newMetronome.observeIsInitialized { initialized in
                self.isInitialized = initialized
                if initialized {
                    self.status = "Ready"
                }
            }
            beatTask = newMetronome.observeCurrentBeat { beat in
                self.currentBeat = beat
            }
            playingTask = newMetronome.observeIsPlaying { playing in
                self.isRunning = playing
            }
        }
    }

    private func reinitializeMetronome(withBeats newBeats: Int32) {
        guard let samaPath = samaSamplePath,
              let beatPath = beatSamplePath else {
            return
        }

        Task {
            // Cancel existing observers
            beatTask?.cancel()
            playingTask?.cancel()
            initializedTask?.cancel()

            // Release old metronome
            metronome?.release()

            await MainActor.run {
                isInitialized = false
                status = "Reinitializing..."
            }

            // CREATE NEW METRONOME - ZERO-CONFIG FACTORY!
            let newMetronome = SonixMetronome.create(
                samaSamplePath: samaPath,
                beatSamplePath: beatPath,
                bpm: bpm,
                beatsPerCycle: newBeats
            )

            await MainActor.run {
                metronome = newMetronome
                currentBeat = 0

                // Type-safe observers - no force casts needed!
                initializedTask = newMetronome.observeIsInitialized { initialized in
                    self.isInitialized = initialized
                    if initialized {
                        self.status = "Ready"
                    }
                }
                beatTask = newMetronome.observeCurrentBeat { beat in
                    self.currentBeat = beat
                }
                playingTask = newMetronome.observeIsPlaying { playing in
                    self.isRunning = playing
                }
            }
        }
    }

    @ViewBuilder
    private func beatsButton(count: Int) -> some View {
        if count == Int(beatsPerCycle) {
            Button("\(count)") {
                if beatsPerCycle != Int32(count) {
                    beatsPerCycle = Int32(count)
                    reinitializeMetronome(withBeats: Int32(count))
                }
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.small)
            .disabled(isRunning)
        } else {
            Button("\(count)") {
                if beatsPerCycle != Int32(count) {
                    beatsPerCycle = Int32(count)
                    reinitializeMetronome(withBeats: Int32(count))
                }
            }
            .buttonStyle(.bordered)
            .controlSize(.small)
            .disabled(isRunning)
        }
    }

    private func beatColor(isCurrentBeat: Bool, isSama: Bool) -> Color {
        if isCurrentBeat && isSama {
            return .red
        } else if isCurrentBeat {
            return .green
        } else if isSama {
            return .red.opacity(0.3)
        } else {
            return .gray.opacity(0.3)
        }
    }
}

#Preview {
    MetronomeSectionSimplified()
        .padding()
}
